File: Leetcode/Python3/121. Best Time to Buy and Sell Stock/solution.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        propit: int = 0
        min_price: int = prices[0]
        for price in prices:
            if price > min_price:
                propit = max(propit, price - min_price)
            else:
                min_price = price
        return propit


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp = [0] * len(prices)
        buy_price = prices[0]
        for idx in range(1, len(prices)):
            dp[idx] = max(prices[idx] - buy_price, dp[idx - 1])
            if buy_price > prices[idx]:
                buy_price = prices[idx]
        return dp[-1]

File: Leetcode/C++/121. Best Time to Buy and Sell Stock/solution.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/submissions/904508417/

#include <vector>

using namespace std;

class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int cur_stock = prices[0];
        int profits = 0;
        for (int i = 1; i < prices.size(); i++)
        {
            if (cur_stock < prices[i])
            {
                profits = max(profits, prices[i] - cur_stock);
            }
            else
            {
                cur_stock = prices[i];
            }
        }
        return profits;
    }
};